// streamr-client-protocol/dist/src/protocol/tracker_layer/instruction_message/InstructionMessage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const validations_1 = require("../../../utils/validations");
const TrackerMessage_1 = __importDefault(require("../TrackerMessage"));
const SPID_1 = require("../../../utils/SPID");
class InstructionMessage extends TrackerMessage_1.default {
    constructor({ version = TrackerMessage_1.default.LATEST_VERSION, requestId, streamId, streamPartition, nodeIds, counter }) {
        super(version, TrackerMessage_1.default.TYPES.InstructionMessage, requestId);
        (0, validations_1.validateIsNotEmptyString)('streamId', streamId);
        (0, validations_1.validateIsNotNegativeInteger)('streamPartition', streamPartition);
        (0, validations_1.validateIsArray)('nodeIds', nodeIds);
        (0, validations_1.validateIsNotNegativeInteger)('counter', counter);
        this.streamId = streamId;
        this.streamPartition = streamPartition;
        this.nodeIds = nodeIds;
        this.counter = counter;
    }
    getSPID() {
        return SPID_1.SPID.from(this);
    }
}
exports.default = InstructionMessage;
