// @aws-sdk/client-sns/dist-es/models/models_0.js
import { __assign } from "tslib";
export var AddPermissionInput;
(function (AddPermissionInput) {
    AddPermissionInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AddPermissionInput || (AddPermissionInput = {}));
export var CheckIfPhoneNumberIsOptedOutInput;
(function (CheckIfPhoneNumberIsOptedOutInput) {
    CheckIfPhoneNumberIsOptedOutInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CheckIfPhoneNumberIsOptedOutInput || (CheckIfPhoneNumberIsOptedOutInput = {}));
export var CheckIfPhoneNumberIsOptedOutResponse;
(function (CheckIfPhoneNumberIsOptedOutResponse) {
    CheckIfPhoneNumberIsOptedOutResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CheckIfPhoneNumberIsOptedOutResponse || (CheckIfPhoneNumberIsOptedOutResponse = {}));
export var ConfirmSubscriptionInput;
(function (ConfirmSubscriptionInput) {
    ConfirmSubscriptionInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfirmSubscriptionInput || (ConfirmSubscriptionInput = {}));
export var ConfirmSubscriptionResponse;
(function (ConfirmSubscriptionResponse) {
    ConfirmSubscriptionResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfirmSubscriptionResponse || (ConfirmSubscriptionResponse = {}));
export var CreatePlatformApplicationInput;
(function (CreatePlatformApplicationInput) {
    CreatePlatformApplicationInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreatePlatformApplicationInput || (CreatePlatformApplicationInput = {}));
export var CreatePlatformApplicationResponse;
(function (CreatePlatformApplicationResponse) {
    CreatePlatformApplicationResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreatePlatformApplicationResponse || (CreatePlatformApplicationResponse = {}));
export var CreateEndpointResponse;
(function (CreateEndpointResponse) {
    CreateEndpointResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateEndpointResponse || (CreateEndpointResponse = {}));
export var CreatePlatformEndpointInput;
(function (CreatePlatformEndpointInput) {
    CreatePlatformEndpointInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreatePlatformEndpointInput || (CreatePlatformEndpointInput = {}));
export var LanguageCodeString;
(function (LanguageCodeString) {
    LanguageCodeString["de_DE"] = "de-DE";
    LanguageCodeString["en_GB"] = "en-GB";
    LanguageCodeString["en_US"] = "en-US";
    LanguageCodeString["es_419"] = "es-419";
    LanguageCodeString["es_ES"] = "es-ES";
    LanguageCodeString["fr_CA"] = "fr-CA";
    LanguageCodeString["fr_FR"] = "fr-FR";
    LanguageCodeString["it_IT"] = "it-IT";
    LanguageCodeString["jp_JP"] = "ja-JP";
    LanguageCodeString["kr_KR"] = "kr-KR";
    LanguageCodeString["pt_BR"] = "pt-BR";
    LanguageCodeString["zh_CN"] = "zh-CN";
    LanguageCodeString["zh_TW"] = "zh-TW";
})(LanguageCodeString || (LanguageCodeString = {}));
export var CreateSMSSandboxPhoneNumberInput;
(function (CreateSMSSandboxPhoneNumberInput) {
    CreateSMSSandboxPhoneNumberInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateSMSSandboxPhoneNumberInput || (CreateSMSSandboxPhoneNumberInput = {}));
export var CreateSMSSandboxPhoneNumberResult;
(function (CreateSMSSandboxPhoneNumberResult) {
    CreateSMSSandboxPhoneNumberResult.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateSMSSandboxPhoneNumberResult || (CreateSMSSandboxPhoneNumberResult = {}));
export var Tag;
(function (Tag) {
    Tag.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Tag || (Tag = {}));
export var CreateTopicInput;
(function (CreateTopicInput) {
    CreateTopicInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateTopicInput || (CreateTopicInput = {}));
export var CreateTopicResponse;
(function (CreateTopicResponse) {
    CreateTopicResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateTopicResponse || (CreateTopicResponse = {}));
export var DeleteEndpointInput;
(function (DeleteEndpointInput) {
    DeleteEndpointInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteEndpointInput || (DeleteEndpointInput = {}));
export var DeletePlatformApplicationInput;
(function (DeletePlatformApplicationInput) {
    DeletePlatformApplicationInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeletePlatformApplicationInput || (DeletePlatformApplicationInput = {}));
export var DeleteSMSSandboxPhoneNumberInput;
(function (DeleteSMSSandboxPhoneNumberInput) {
    DeleteSMSSandboxPhoneNumberInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteSMSSandboxPhoneNumberInput || (DeleteSMSSandboxPhoneNumberInput = {}));
export var DeleteSMSSandboxPhoneNumberResult;
(function (DeleteSMSSandboxPhoneNumberResult) {
    DeleteSMSSandboxPhoneNumberResult.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteSMSSandboxPhoneNumberResult || (DeleteSMSSandboxPhoneNumberResult = {}));
export var DeleteTopicInput;
(function (DeleteTopicInput) {
    DeleteTopicInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteTopicInput || (DeleteTopicInput = {}));
export var GetEndpointAttributesInput;
(function (GetEndpointAttributesInput) {
    GetEndpointAttributesInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetEndpointAttributesInput || (GetEndpointAttributesInput = {}));
export var GetEndpointAttributesResponse;
(function (GetEndpointAttributesResponse) {
    GetEndpointAttributesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetEndpointAttributesResponse || (GetEndpointAttributesResponse = {}));
export var GetPlatformApplicationAttributesInput;
(function (GetPlatformApplicationAttributesInput) {
    GetPlatformApplicationAttributesInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetPlatformApplicationAttributesInput || (GetPlatformApplicationAttributesInput = {}));
export var GetPlatformApplicationAttributesResponse;
(function (GetPlatformApplicationAttributesResponse) {
    GetPlatformApplicationAttributesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetPlatformApplicationAttributesResponse || (GetPlatformApplicationAttributesResponse = {}));
export var GetSMSAttributesInput;
(function (GetSMSAttributesInput) {
    GetSMSAttributesInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetSMSAttributesInput || (GetSMSAttributesInput = {}));
export var GetSMSAttributesResponse;
(function (GetSMSAttributesResponse) {
    GetSMSAttributesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetSMSAttributesResponse || (GetSMSAttributesResponse = {}));
export var GetSMSSandboxAccountStatusInput;
(function (GetSMSSandboxAccountStatusInput) {
    GetSMSSandboxAccountStatusInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetSMSSandboxAccountStatusInput || (GetSMSSandboxAccountStatusInput = {}));
export var GetSMSSandboxAccountStatusResult;
(function (GetSMSSandboxAccountStatusResult) {
    GetSMSSandboxAccountStatusResult.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetSMSSandboxAccountStatusResult || (GetSMSSandboxAccountStatusResult = {}));
export var GetSubscriptionAttributesInput;
(function (GetSubscriptionAttributesInput) {
    GetSubscriptionAttributesInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetSubscriptionAttributesInput || (GetSubscriptionAttributesInput = {}));
export var GetSubscriptionAttributesResponse;
(function (GetSubscriptionAttributesResponse) {
    GetSubscriptionAttributesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetSubscriptionAttributesResponse || (GetSubscriptionAttributesResponse = {}));
export var GetTopicAttributesInput;
(function (GetTopicAttributesInput) {
    GetTopicAttributesInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetTopicAttributesInput || (GetTopicAttributesInput = {}));
export var GetTopicAttributesResponse;
(function (GetTopicAttributesResponse) {
    GetTopicAttributesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetTopicAttributesResponse || (GetTopicAttributesResponse = {}));
export var ListEndpointsByPlatformApplicationInput;
(function (ListEndpointsByPlatformApplicationInput) {
    ListEndpointsByPlatformApplicationInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListEndpointsByPlatformApplicationInput || (ListEndpointsByPlatformApplicationInput = {}));
export var Endpoint;
(function (Endpoint) {
    Endpoint.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Endpoint || (Endpoint = {}));
export var ListEndpointsByPlatformApplicationResponse;
(function (ListEndpointsByPlatformApplicationResponse) {
    ListEndpointsByPlatformApplicationResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListEndpointsByPlatformApplicationResponse || (ListEndpointsByPlatformApplicationResponse = {}));
export var ListOriginationNumbersRequest;
(function (ListOriginationNumbersRequest) {
    ListOriginationNumbersRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListOriginationNumbersRequest || (ListOriginationNumbersRequest = {}));
export var NumberCapability;
(function (NumberCapability) {
    NumberCapability["MMS"] = "MMS";
    NumberCapability["SMS"] = "SMS";
    NumberCapability["VOICE"] = "VOICE";
})(NumberCapability || (NumberCapability = {}));
export var RouteType;
(function (RouteType) {
    RouteType["Premium"] = "Premium";
    RouteType["Promotional"] = "Promotional";
    RouteType["Transactional"] = "Transactional";
})(RouteType || (RouteType = {}));
export var PhoneNumberInformation;
(function (PhoneNumberInformation) {
    PhoneNumberInformation.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PhoneNumberInformation || (PhoneNumberInformation = {}));
export var ListOriginationNumbersResult;
(function (ListOriginationNumbersResult) {
    ListOriginationNumbersResult.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListOriginationNumbersResult || (ListOriginationNumbersResult = {}));
export var ListPhoneNumbersOptedOutInput;
(function (ListPhoneNumbersOptedOutInput) {
    ListPhoneNumbersOptedOutInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListPhoneNumbersOptedOutInput || (ListPhoneNumbersOptedOutInput = {}));
export var ListPhoneNumbersOptedOutResponse;
(function (ListPhoneNumbersOptedOutResponse) {
    ListPhoneNumbersOptedOutResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListPhoneNumbersOptedOutResponse || (ListPhoneNumbersOptedOutResponse = {}));
export var ListPlatformApplicationsInput;
(function (ListPlatformApplicationsInput) {
    ListPlatformApplicationsInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListPlatformApplicationsInput || (ListPlatformApplicationsInput = {}));
export var PlatformApplication;
(function (PlatformApplication) {
    PlatformApplication.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PlatformApplication || (PlatformApplication = {}));
export var ListPlatformApplicationsResponse;
(function (ListPlatformApplicationsResponse) {
    ListPlatformApplicationsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListPlatformApplicationsResponse || (ListPlatformApplicationsResponse = {}));
export var ListSMSSandboxPhoneNumbersInput;
(function (ListSMSSandboxPhoneNumbersInput) {
    ListSMSSandboxPhoneNumbersInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListSMSSandboxPhoneNumbersInput || (ListSMSSandboxPhoneNumbersInput = {}));
export var SMSSandboxPhoneNumberVerificationStatus;
(function (SMSSandboxPhoneNumberVerificationStatus) {
    SMSSandboxPhoneNumberVerificationStatus["Pending"] = "Pending";
    SMSSandboxPhoneNumberVerificationStatus["Verified"] = "Verified";
})(SMSSandboxPhoneNumberVerificationStatus || (SMSSandboxPhoneNumberVerificationStatus = {}));
export var SMSSandboxPhoneNumber;
(function (SMSSandboxPhoneNumber) {
    SMSSandboxPhoneNumber.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SMSSandboxPhoneNumber || (SMSSandboxPhoneNumber = {}));
export var ListSMSSandboxPhoneNumbersResult;
(function (ListSMSSandboxPhoneNumbersResult) {
    ListSMSSandboxPhoneNumbersResult.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListSMSSandboxPhoneNumbersResult || (ListSMSSandboxPhoneNumbersResult = {}));
export var ListSubscriptionsInput;
(function (ListSubscriptionsInput) {
    ListSubscriptionsInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListSubscriptionsInput || (ListSubscriptionsInput = {}));
export var Subscription;
(function (Subscription) {
    Subscription.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Subscription || (Subscription = {}));
export var ListSubscriptionsResponse;
(function (ListSubscriptionsResponse) {
    ListSubscriptionsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListSubscriptionsResponse || (ListSubscriptionsResponse = {}));
export var ListSubscriptionsByTopicInput;
(function (ListSubscriptionsByTopicInput) {
    ListSubscriptionsByTopicInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListSubscriptionsByTopicInput || (ListSubscriptionsByTopicInput = {}));
export var ListSubscriptionsByTopicResponse;
(function (ListSubscriptionsByTopicResponse) {
    ListSubscriptionsByTopicResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListSubscriptionsByTopicResponse || (ListSubscriptionsByTopicResponse = {}));
export var ListTagsForResourceRequest;
(function (ListTagsForResourceRequest) {
    ListTagsForResourceRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListTagsForResourceRequest || (ListTagsForResourceRequest = {}));
export var ListTagsForResourceResponse;
(function (ListTagsForResourceResponse) {
    ListTagsForResourceResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListTagsForResourceResponse || (ListTagsForResourceResponse = {}));
export var ListTopicsInput;
(function (ListTopicsInput) {
    ListTopicsInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListTopicsInput || (ListTopicsInput = {}));
export var Topic;
(function (Topic) {
    Topic.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Topic || (Topic = {}));
export var ListTopicsResponse;
(function (ListTopicsResponse) {
    ListTopicsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListTopicsResponse || (ListTopicsResponse = {}));
export var OptInPhoneNumberInput;
(function (OptInPhoneNumberInput) {
    OptInPhoneNumberInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(OptInPhoneNumberInput || (OptInPhoneNumberInput = {}));
export var OptInPhoneNumberResponse;
(function (OptInPhoneNumberResponse) {
    OptInPhoneNumberResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(OptInPhoneNumberResponse || (OptInPhoneNumberResponse = {}));
export var MessageAttributeValue;
(function (MessageAttributeValue) {
    MessageAttributeValue.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(MessageAttributeValue || (MessageAttributeValue = {}));
export var PublishInput;
(function (PublishInput) {
    PublishInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PublishInput || (PublishInput = {}));
export var PublishResponse;
(function (PublishResponse) {
    PublishResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PublishResponse || (PublishResponse = {}));
export var PublishBatchRequestEntry;
(function (PublishBatchRequestEntry) {
    PublishBatchRequestEntry.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PublishBatchRequestEntry || (PublishBatchRequestEntry = {}));
export var PublishBatchInput;
(function (PublishBatchInput) {
    PublishBatchInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PublishBatchInput || (PublishBatchInput = {}));
export var BatchResultErrorEntry;
(function (BatchResultErrorEntry) {
    BatchResultErrorEntry.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(BatchResultErrorEntry || (BatchResultErrorEntry = {}));
export var PublishBatchResultEntry;
(function (PublishBatchResultEntry) {
    PublishBatchResultEntry.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PublishBatchResultEntry || (PublishBatchResultEntry = {}));
export var PublishBatchResponse;
(function (PublishBatchResponse) {
    PublishBatchResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PublishBatchResponse || (PublishBatchResponse = {}));
export var RemovePermissionInput;
(function (RemovePermissionInput) {
    RemovePermissionInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(RemovePermissionInput || (RemovePermissionInput = {}));
export var SetEndpointAttributesInput;
(function (SetEndpointAttributesInput) {
    SetEndpointAttributesInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetEndpointAttributesInput || (SetEndpointAttributesInput = {}));
export var SetPlatformApplicationAttributesInput;
(function (SetPlatformApplicationAttributesInput) {
    SetPlatformApplicationAttributesInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetPlatformApplicationAttributesInput || (SetPlatformApplicationAttributesInput = {}));
export var SetSMSAttributesInput;
(function (SetSMSAttributesInput) {
    SetSMSAttributesInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetSMSAttributesInput || (SetSMSAttributesInput = {}));
export var SetSMSAttributesResponse;
(function (SetSMSAttributesResponse) {
    SetSMSAttributesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetSMSAttributesResponse || (SetSMSAttributesResponse = {}));
export var SetSubscriptionAttributesInput;
(function (SetSubscriptionAttributesInput) {
    SetSubscriptionAttributesInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetSubscriptionAttributesInput || (SetSubscriptionAttributesInput = {}));
export var SetTopicAttributesInput;
(function (SetTopicAttributesInput) {
    SetTopicAttributesInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetTopicAttributesInput || (SetTopicAttributesInput = {}));
export var SubscribeInput;
(function (SubscribeInput) {
    SubscribeInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SubscribeInput || (SubscribeInput = {}));
export var SubscribeResponse;
(function (SubscribeResponse) {
    SubscribeResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SubscribeResponse || (SubscribeResponse = {}));
export var TagResourceRequest;
(function (TagResourceRequest) {
    TagResourceRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(TagResourceRequest || (TagResourceRequest = {}));
export var TagResourceResponse;
(function (TagResourceResponse) {
    TagResourceResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(TagResourceResponse || (TagResourceResponse = {}));
export var UnsubscribeInput;
(function (UnsubscribeInput) {
    UnsubscribeInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UnsubscribeInput || (UnsubscribeInput = {}));
export var UntagResourceRequest;
(function (UntagResourceRequest) {
    UntagResourceRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UntagResourceRequest || (UntagResourceRequest = {}));
export var UntagResourceResponse;
(function (UntagResourceResponse) {
    UntagResourceResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UntagResourceResponse || (UntagResourceResponse = {}));
export var VerifySMSSandboxPhoneNumberInput;
(function (VerifySMSSandboxPhoneNumberInput) {
    VerifySMSSandboxPhoneNumberInput.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(VerifySMSSandboxPhoneNumberInput || (VerifySMSSandboxPhoneNumberInput = {}));
export var VerifySMSSandboxPhoneNumberResult;
(function (VerifySMSSandboxPhoneNumberResult) {
    VerifySMSSandboxPhoneNumberResult.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(VerifySMSSandboxPhoneNumberResult || (VerifySMSSandboxPhoneNumberResult = {}));
